Mathlib/Geometry/Euclidean/Similarity.lean

```diff
@@ -0,0 +1,278 @@
+    (i j : Fin 3) :
+    ∃ r : ℝ, r > 0 ∧ dist (t₁ i) (t₁ j) = r * dist (t₂ i) (t₂ j) := by
+  rw [triangle_similar_iff_exists_dist_eq] at h
+  rcases h with ⟨r, hr_pos, hdist⟩
+  use r
+  exact ⟨hr_pos, hdist i j⟩
+
+/-- For two similar triangles, the corresponding angles are equal. -/
+theorem angle_eq_of_similar (h : ![a, b, c] ∼ ![a', b', c']) :
+    ∠ a b c = ∠ a' b' c' := by
+  rw [triangle_similar_iff_exists_dist_eq] at h
+  rcases h with ⟨r, hr_pos, hdist⟩
+  have h_ab : dist a b = r * dist a' b' := hdist 0 1
+  have h_cb : dist c b = r * dist c' b' := hdist 2 1
+  have h_ac : dist a c = r * dist a' c' := hdist 0 2
+  have h_cos := law_cos a b c
+  rw [h_ab, h_cb, h_ac] at h_cos
+  field_simp at h_cos
+  have h_cos' := law_cos a' b' c'
+  field_simp at h_cos'
+  rw [h_cos', sub_right_inj] at h_cos
+  by_cases heq : dist a' b' * dist c' b' * 2 = 0
+  · rw [mul_eq_zero_iff_right (by norm_num), mul_eq_zero] at heq
+    rcases heq with h1 | h2
+    · have h_dist_ab : dist a b = 0 := by
+        rw [h_ab, h1, mul_zero]
+      rw [dist_eq_zero] at h_dist_ab h1
+      simp_rw [h_dist_ab, h1, angle_self_left]
+    · have h_dist_cb : dist c b = 0 := by
+        rw [h_cb, h2, mul_zero]
+      rw [dist_eq_zero] at h_dist_cb h2
+      simp_rw [h_dist_cb, h2, angle_self_right]
+  rw [mul_right_inj' heq] at h_cos
+  apply Real.injOn_cos at h_cos
+  · rw [h_cos]
+  repeat exact ⟨angle_nonneg _ _ _, angle_le_pi _ _ _ ⟩
+
+/-- Reindexing both triangles by the same permutation preserves similarity. -/
+theorem similar_reindex {t₁ : ι → P₁} {t₂ : ι → P₂} (e : Equiv.Perm (ι)) :
+    t₁ ∼ t₂ ↔ (t₁ ∘ e) ∼ (t₂ ∘ e) := by
+  simp_rw [similar_iff_exists_dist_eq]
+  constructor
+  · rintro ⟨r, hr0, hdist⟩
+    refine ⟨r, hr0, ?_⟩
+    intro i j
+    simpa using hdist (e i) (e j)
+  · rintro ⟨r, hr0, hdist⟩
+    refine ⟨r, hr0, ?_⟩
+    intro i j
+    simpa using hdist (e.symm i) (e.symm j)
+
+theorem similar_comm_left (h : ![a, b, c] ∼ ![a', b', c']) :
+    ![b, a, c] ∼ ![b', a', c'] := by
+  have hl : ![b, a, c] = ![a, b, c] ∘ Equiv.swap 0 1 := by
+    ext i
+    fin_cases i <;> simp; rfl
+  have hr : ![b', a', c'] = ![a', b', c'] ∘ Equiv.swap 0 1 := by
+    ext i
+    fin_cases i <;> simp; rfl
+  grind [similar_reindex]
+
+theorem similar_comm_right (h : ![a, b, c] ∼ ![a', b', c']) :
+    ![a, c, b] ∼ ![a', c', b'] := by
+  have hl : ![a, c, b] = ![a, b, c] ∘ Equiv.swap 1 2 := by
+    ext i
+    fin_cases i <;> simp; rfl
+  have hr : ![a', c', b'] = ![a', b', c'] ∘ Equiv.swap 1 2 := by
+    ext i
+    fin_cases i <;> simp; rfl
+  grind [similar_reindex]
+
+theorem similar_reverse (h : ![a, b, c] ∼ ![a', b', c']) :
+    ![c, b, a] ∼ ![c', b', a'] := by
+  have hl : ![c, b, a] = ![a, b, c] ∘ Equiv.swap 0 2 := by
+    ext i
+    fin_cases i <;> simp; rfl
+  have hr : ![c', b', a'] = ![a', b', c'] ∘ Equiv.swap 0 2 := by
+    ext i
+    fin_cases i <;> simp; rfl
+  grind [similar_reindex]
```

Comment:
All three of these should work in the generic file for an arbitrary `PseudoEMetricSpace`.